### PR TITLE
Attachments

### PR DESCRIPTION
**Pull request by kbatbouta**
## Changes

- Added a complete system for attachments. This system mirror the way apparel work but use it for weapons. The system is modular and allow us a lot of flexibility.
- Added a crafting system for modifying weapons.
- Added my UI lib `RocketGUI` which allowed us to make the new weapon attachment editor window and debug window.
- Added stats for most major properties related to weapons. which allowed us to implement a variable size magazine, variable fire rate, variable amount of bullets per burst and much more.
- `ToolCE` can now specify an attachment as a requirement. This is the system required for bayonets.
- Added attachment options use in pawn gen.

### CODE REVIEW

File: Source/CombatExtended/Harmony/Harmony_PawnRenderer.cs [hunk 1]
```diff
@@ -2,6 +2,7 @@
 using System.Collections;
 using System.Collections.Generic;
 using System.Linq;
+using System.Net.Mail;
```
Comment (Alicecomma) [inline Source/CombatExtended/Harmony/Harmony_PawnRenderer.cs:5]: "

File: Defs/ThingDefs_Misc/Weapons_Ranged.xml [hunk 2]
```diff
@@ -251,37 +251,530 @@
...
         <DrawOffset>0.0,0.0</DrawOffset>
       </li>
-    -->  
-    </modExtensions>	
+    -->
+    </modExtensions>
   </ThingDef>
 
+
+  <CombatExtended.WeaponPlatformDef ParentName="BaseMakeableGun">
```
Comment (Alicecomma) [inline Defs/ThingDefs_Misc/Weapons_Ranged.xml:288]: Move to different file on full release

File: Source/CombatExtended/CombatExtended/CE_Utility.cs [hunk 3]
```diff
@@ -26,6 +26,138 @@ public static float CameraAltitude
...
+                    continue;
+                // stop since we found an override modifier.
+                val = modifier.value;                
+                return;
+            }
+            for (int i = 0;i < links.Count; i++)
+            {
+                AttachmentLink link = links[i];               
+                StatModifier modifier = link.statOffsets?.FirstOrFallback(m => m.stat == stat, null) ?? null;
+                if (modifier == null)
```
Comment (Alicecomma) [inline Source/CombatExtended/CombatExtended/CE_Utility.cs:145]: Or modifier == 0

File: Source/CombatExtended/CombatExtended/CE_Utility.cs [hunk 4]
```diff
@@ -26,6 +26,138 @@ public static float CameraAltitude
...
+                    continue;
+                val += modifier.value;
+            }
+            for (int i = 0; i < links.Count; i++)
+            {
+                AttachmentLink link = links[i];                
+                StatModifier modifier = link.statMultipliers?.FirstOrFallback(m => m.stat == stat, null) ?? null;
+                if (modifier == null || modifier.value == 0)
```
Comment (Alicecomma) [inline Source/CombatExtended/CombatExtended/CE_Utility.cs:153]: Or modifier == 1 or modifier < 0

File: Source/CombatExtended/CombatExtended/DefOfs/CE_ThingCategoryDefOf.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using RimWorld;
+using Verse;
+
+namespace CombatExtended
+{
+    //[DefOf]
```
Comment (Alicecomma) [inline Source/CombatExtended/CombatExtended/DefOfs/CE_ThingCategoryDefOf.cs:7]: Empty

File: Source/CombatExtended/CombatExtended/StatParts/StatPart_Attachments.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
+﻿using System;
+using System.Linq;
+using RimWorld;
+
+namespace CombatExtended
+{
+    public class StatPart_Attachments : StatPart
+    {
+        public StatPart_Attachments()
+        {
```
Comment (Alicecomma) [inline Source/CombatExtended/CombatExtended/StatParts/StatPart_Attachments.cs:10]: Needs explanations in release

File: Source/CombatExtended/CombatExtended/Window_Main.cs [hunk 7]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using UnityEngine;
+using Verse;
+
+namespace CombatExtended
+{
+    public class Window_Main : Window
```
Comment (Alicecomma) [inline Source/CombatExtended/CombatExtended/Window_Main.cs:7]: Empty

File: Source/CombatExtended/Harmony/Harmony_PawnRenderer.cs (hunk 1 above)
Comment (Alicecomma) [reply Source/CombatExtended/Harmony/Harmony_PawnRenderer.cs:5]: (Mail is added automatically when 'Attachment' is used I guess)

File: Defs/ThingDefs_Misc/Weapons_Ranged.xml (hunk 2 above)
Comment (androsshi) [reply Defs/ThingDefs_Misc/Weapons_Ranged.xml:288]: I plan to remove / comment these after it's merged

**Review by Alicecomma — APPROVED** [review]

**Review by androsshi — APPROVED** [review]

File: Source/CombatExtended/CombatExtended/Jobs/WorkGiver_ModifyWeapon.cs [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using RimWorld;
+using UnityEngine;
+using Verse;
+using Verse.AI;
+
+namespace CombatExtended
+{
+    public class WorkGiver_ModifyWeapon : WorkGiver_Scanner
```
Comment (Alicecomma) [inline Source/CombatExtended/CombatExtended/Jobs/WorkGiver_ModifyWeapon.cs:11]: Please add a method like `// RimWorld.WorkGiver_DoBill private Job StartOrResumeBillJob(Pawn pawn, IBillGiver giver)` to send ticker job fail reasons like `JobFailReason.Is("UnderRequiredSkill".Translate(skillRequirement.minLevel), bill.Label);`  for lacking the right ingredients on the map.

**Review by Alicecomma — CHANGES_REQUESTED** [review]

File: Source/CombatExtended/CombatExtended/Jobs/WorkGiver_ModifyWeapon.cs (hunk 8 above)
Comment (kbatbouta) [reply Source/CombatExtended/CombatExtended/Jobs/WorkGiver_ModifyWeapon.cs:11]: I've never done this before and i don't see any reason to do it now.

**Review by Alicecomma — APPROVED** [review]

**Review by androsshi — APPROVED** [review]

**Pull request merged**